File: server/src/test/java/rikka/shizuku/server/util/PackageIdentityTest.java
package rikka.shizuku.server.util;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;
import static org.mockito.ArgumentMatchers.anyInt;
import static org.mockito.ArgumentMatchers.anyLong;
import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

import android.content.pm.ApplicationInfo;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;
import android.content.pm.Signature;
import android.content.pm.SigningInfo;

import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.MockedStatic;
import org.mockito.Mockito;
import org.robolectric.RobolectricTestRunner;
import org.robolectric.annotation.Config;

import java.lang.reflect.Field;

import rikka.hidden.compat.PackageManagerApis;

/** Three-way classification, and the signer comparison the identity model rests on. */
@RunWith(RobolectricTestRunner.class)
@Config(sdk = 34, manifest = Config.NONE)
public class PackageIdentityTest {

    static final String PACKAGE = "eu.darken.porter.probe";
    static final int UID = 10123;

    static final Signature ORIGINAL = new Signature("0a0b");
    static final Signature ROTATED = new Signature("0c0d");
    static final Signature FOREIGN = new Signature("0e0f");
    static final Signature SECOND = new Signature("1a1b");

    private MockedStatic<PackageManagerApis> packages;

    @Before
    public void setup() throws Exception {
        packages = Mockito.mockStatic(PackageManagerApis.class);
        compatField("sPackageManager", null);
        compatField("sGetPackageInfoMethod", null);
    }

    @After
    public void teardown() {
        packages.close();
    }

    private static void compatField(String name, Object value) throws Exception {
        Field field = Android17Compat.class.getDeclaredField(name);
        field.setAccessible(true);
        field.set(null, value);
    }

    /** A package installed for one user, signed by {@code signers} with lineage {@code history}. */
    static PackageInfo installed(Signature[] signers, Signature[] history) {
        PackageInfo packageInfo = new PackageInfo();
        packageInfo.packageName = PACKAGE;
        packageInfo.applicationInfo = new ApplicationInfo();
        packageInfo.applicationInfo.uid = UID;
        packageInfo.applicationInfo.flags = ApplicationInfo.FLAG_INSTALLED;
        SigningInfo signingInfo = mock(SigningInfo.class);
        when(signingInfo.hasMultipleSigners()).thenReturn(signers.length > 1);
        when(signingInfo.getApkContentsSigners()).thenReturn(signers);
        when(signingInfo.getSigningCertificateHistory()).thenReturn(signers.length > 1 ? null : history);
        packageInfo.signingInfo = signingInfo;
        return packageInfo;
    }

    static PackageIdentity.Identity identityOf(Signature... signers) {
        return PackageIdentity.identityOf(installed(signers, signers));
    }

    static PackageIdentity.Observed observe(PackageInfo packageInfo, int userId) {
        PackageIdentity.Result result = PackageIdentity.classify(packageInfo, userId);
        assertEquals(PackageIdentity.State.PRESENT, result.state);
        return result.observed;
    }

    @Test
    public void aHiddenPackageIsStillInstalled() {
        // pm hide leaves the installed bit set, and MATCH_UNINSTALLED_PACKAGES still returns it.
        PackageInfo hidden = installed(new Signature[]{ORIGINAL}, new Signature[]{ORIGINAL});
        assertEquals(PackageIdentity.State.PRESENT, PackageIdentity.classify(hidden, 0).state);
    }

    @Test
    public void aRetainedDataResidueIsAbsent() {
        PackageInfo residue = installed(new Signature[]{ORIGINAL}, new Signature[]{ORIGINAL});
        residue.applicationInfo.flags &= ~ApplicationInfo.FLAG_INSTALLED;
        assertEquals(PackageIdentity.State.ABSENT, PackageIdentity.classify(residue, 0).state);
    }

    @Test
    public void aMissingPackageIsAbsent() {
        assertEquals(PackageIdentity.State.ABSENT, PackageIdentity.classify(null, 0).state);
    }

    @Test
    public void anAnswerWithoutSignersIsAFailureRatherThanEvidence() {
        PackageInfo unreadable = installed(new Signature[]{ORIGINAL}, new Signature[]{ORIGINAL});
        when(unreadable.signingInfo.getApkContentsSigners()).thenReturn(new Signature[0]);
        assertEquals(PackageIdentity.State.LOOKUP_FAILED, PackageIdentity.classify(unreadable, 0).state);
    }

    @Test
    public void aFailedLookupIsNeverReportedAsAbsence() {
        packages.when(() -> PackageManagerApis.getPackageInfo(anyString(), anyLong(), anyInt()))
                .thenThrow(new SecurityException("Shell cannot query across users"));

        PackageIdentity.Result result = PackageIdentity.of(PACKAGE, 0);
        assertEquals(PackageIdentity.State.LOOKUP_FAILED, result.state);
        assertTrue(result.cause instanceof SecurityException);
    }

    @Test
    public void aMissingHiddenMethodAlsoReportsFailureWhenTheFallbackCannotAnswer() {
        packages.when(() -> PackageManagerApis.getPackageInfo(anyString(), anyLong(), anyInt()))
                .thenThrow(new NoSuchMethodError("getPackageInfo"));

        assertEquals(PackageIdentity.State.LOOKUP_FAILED, PackageIdentity.of(PACKAGE, 0).state);
    }

    @Test
    public void reconciliationStillWorksThroughTheReflectionFallback() throws Exception {
        packages.when(() -> PackageManagerApis.getPackageInfo(anyString(), anyLong(), anyInt()))
                .thenThrow(new NoSuchMethodError("getPackageInfo"));
        Android17CompatTest.FakePackageManager fake = new Android17CompatTest.FakePackageManager();
        fake.answer = installed(new Signature[]{ORIGINAL}, new Signature[]{ORIGINAL});
        compatField("sPackageManager", fake);

        PackageIdentity.Result result = PackageIdentity.of(PACKAGE, 10);
        assertEquals(PackageIdentity.State.PRESENT, result.state);
        assertEquals(UID, result.observed.appId);
    }

    @Test
    public void aWorkingLookupIsPresentWithTheCallersAppId() {
        PackageInfo present = installed(new Signature[]{ORIGINAL}, new Signature[]{ORIGINAL});
        packages.when(() -> PackageManagerApis.getPackageInfo(anyString(), anyLong(), anyInt()))
                .thenReturn(present);

        PackageIdentity.Result result = PackageIdentity.of(PACKAGE, 10);
        assertEquals(PackageIdentity.State.PRESENT, result.state);
        assertEquals(UID, result.observed.appId);
        assertEquals(10, result.observed.userId);
    }

    @Test
    public void theLookupFlagsAskForSignaturesAndForUninstalledPackages() {
        long flags = PackageIdentity.lookupFlags();
        assertTrue((flags & PackageManager.MATCH_UNINSTALLED_PACKAGES) != 0);
        assertTrue((flags & PackageManager.GET_SIGNING_CERTIFICATES) != 0);
    }

    @Test
    public void anOrdinaryUpdateBySameSignerMatches() {
        PackageIdentity.Identity recorded = identityOf(ORIGINAL);
        assertTrue(recorded.matches(observe(installed(new Signature[]{ORIGINAL}, new Signature[]{ORIGINAL}), 0)));
    }

    @Test
    public void aRotatedKeyMatchesThroughItsLineage() {
        PackageIdentity.Identity recorded = identityOf(ORIGINAL);
        PackageInfo rotated = installed(new Signature[]{ROTATED}, new Signature[]{ORIGINAL, ROTATED});
        assertTrue(recorded.matches(observe(rotated, 0)));
    }

    @Test
    public void aForeignSignerDoesNotMatch() {
        PackageIdentity.Identity recorded = identityOf(ORIGINAL);
        PackageInfo foreign = installed(new Signature[]{FOREIGN}, new Signature[]{FOREIGN});
        assertFalse(recorded.matches(observe(foreign, 0)));
    }

    @Test
    public void aMultiSignerPackageComparesAsAnUnorderedSet() {
        PackageIdentity.Identity recorded = identityOf(ORIGINAL, SECOND);
        PackageInfo permuted = installed(new Signature[]{SECOND, ORIGINAL}, null);
        assertTrue(recorded.matches(observe(permuted, 0)));
    }

    @Test
    public void aMultiSignerPackageWithAForeignSignerDoesNotMatch() {
        // hasMultipleSigners() means no lineage exists, so the whole set is the identity.
        PackageIdentity.Identity recorded = identityOf(ORIGINAL, SECOND);
        PackageInfo replaced = installed(new Signature[]{ORIGINAL, FOREIGN}, null);
        assertFalse(recorded.matches(observe(replaced, 0)));
    }

    @Test
    public void oneCoSignerOfATwoSignerIdentityIsNotThatIdentity() {
        // Recorded as signed by ORIGINAL and SECOND together. A package signed by ORIGINAL alone,
        // carrying ORIGINAL in its lineage, dropped a signing authority; that is a different
        // installation, not a rotation of the pair.
        PackageIdentity.Identity recorded = identityOf(ORIGINAL, SECOND);
        PackageInfo dropped = installed(new Signature[]{ORIGINAL}, new Signature[]{ORIGINAL});
        assertFalse("a single co-signer matched a two-signer identity",
                recorded.matches(observe(dropped, 0)));
    }

    @Test
    public void aDifferentAppIdDoesNotMatchEvenWithTheSameSigner() {
        PackageIdentity.Identity recorded = identityOf(ORIGINAL);
        PackageInfo moved = installed(new Signature[]{ORIGINAL}, new Signature[]{ORIGINAL});
        moved.applicationInfo.uid = UID + 1;
        assertFalse(recorded.matches(observe(moved, 0)));
    }

    @Test
    public void anUnreadableSignatureDoesNotBecomeAnIdentityWithoutSigners() {
        PackageInfo unreadable = installed(new Signature[]{ORIGINAL}, new Signature[]{ORIGINAL});
        when(unreadable.signingInfo.getApkContentsSigners()).thenReturn(new Signature[0]);
        // The identical reading, through the other entry point, is a failure rather than an answer.
        assertEquals(PackageIdentity.State.LOOKUP_FAILED, PackageIdentity.classify(unreadable, 0).state);

        PackageIdentity.Identity identity = PackageIdentity.identityOf(unreadable);
        assertTrue("identityOf produced an identity with no signers: " + identity,
                identity == null || !identity.signerDigests.isEmpty());
    }

    @Test
    public void twoUsersReportingDifferentSignersIsUnstableRatherThanAReplacement() {
        PackageIdentity.Observed first = observe(installed(new Signature[]{ORIGINAL}, new Signature[]{ORIGINAL}), 0);
        PackageIdentity.Observed second = observe(installed(new Signature[]{FOREIGN}, new Signature[]{FOREIGN}), 10);
        assertFalse(PackageIdentity.sameInstallation(first, second));
        assertTrue(PackageIdentity.sameInstallation(first, first));
    }
}

File: server/src/test/java/eu/darken/porter/privileged/util/PackageIdentityTest.kt
package eu.darken.porter.privileged.util

import android.content.pm.ApplicationInfo
import android.content.pm.PackageInfo
import android.content.pm.PackageManager
import android.content.pm.Signature
import android.content.pm.SigningInfo
import org.junit.After
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.ArgumentMatchers.anyInt
import org.mockito.ArgumentMatchers.anyLong
import org.mockito.ArgumentMatchers.anyString
import org.mockito.MockedStatic
import org.mockito.Mockito
import org.mockito.Mockito.mock
import org.mockito.Mockito.`when`
import org.robolectric.RobolectricTestRunner
import org.robolectric.annotation.Config
import rikka.hidden.compat.PackageManagerApis

/** Three-way classification, and the signer comparison the identity model rests on. */
@RunWith(RobolectricTestRunner::class)
@Config(sdk = [34], manifest = Config.NONE)
class PackageIdentityTest {

    private lateinit var packages: MockedStatic<PackageManagerApis>

    @Before
    fun setup() {
        packages = Mockito.mockStatic(PackageManagerApis::class.java)
        Android17CompatTest.set("sPackageManager", null)
        Android17CompatTest.set("sGetPackageInfoMethod", null)
    }

    @After
    fun teardown() {
        packages.close()
    }

    private fun directCallThrows(failure: Throwable) {
        packages.`when`<PackageInfo?> { PackageManagerApis.getPackageInfo(anyString(), anyLong(), anyInt()) }
            .thenThrow(failure)
    }

    @Test
    fun aHiddenPackageIsStillInstalled() {
        // pm hide leaves the installed bit set, and MATCH_UNINSTALLED_PACKAGES still returns it.
        val hidden = installed(arrayOf(ORIGINAL), arrayOf(ORIGINAL))
        assertEquals(PackageIdentity.State.PRESENT, PackageIdentity.classify(hidden, 0).state)
    }

    @Test
    fun aRetainedDataResidueIsAbsent() {
        val residue = installed(arrayOf(ORIGINAL), arrayOf(ORIGINAL))
        residue.applicationInfo!!.flags = residue.applicationInfo!!.flags and ApplicationInfo.FLAG_INSTALLED.inv()
        assertEquals(PackageIdentity.State.ABSENT, PackageIdentity.classify(residue, 0).state)
    }

    @Test
    fun aMissingPackageIsAbsent() {
        assertEquals(PackageIdentity.State.ABSENT, PackageIdentity.classify(null, 0).state)
    }

    @Test
    fun anAnswerWithoutSignersIsAFailureRatherThanEvidence() {
        val unreadable = installed(arrayOf(ORIGINAL), arrayOf(ORIGINAL))
        `when`(unreadable.signingInfo!!.apkContentsSigners).thenReturn(arrayOf())
        assertEquals(PackageIdentity.State.LOOKUP_FAILED, PackageIdentity.classify(unreadable, 0).state)
    }

    @Test
    fun aFailedLookupIsNeverReportedAsAbsence() {
        directCallThrows(SecurityException("Shell cannot query across users"))

        val result = PackageIdentity.of(PACKAGE, 0)
        assertEquals(PackageIdentity.State.LOOKUP_FAILED, result.state)
        assertTrue(result.cause is SecurityException)
    }

    @Test
    fun aMissingHiddenMethodAlsoReportsFailureWhenTheFallbackCannotAnswer() {
        directCallThrows(NoSuchMethodError("getPackageInfo"))

        assertEquals(PackageIdentity.State.LOOKUP_FAILED, PackageIdentity.of(PACKAGE, 0).state)
    }

    @Test
    fun reconciliationStillWorksThroughTheReflectionFallback() {
        directCallThrows(NoSuchMethodError("getPackageInfo"))
        val fake = Android17CompatTest.FakePackageManager()
        fake.answer = installed(arrayOf(ORIGINAL), arrayOf(ORIGINAL))
        Android17CompatTest.set("sPackageManager", fake)

        val result = PackageIdentity.of(PACKAGE, 10)
        assertEquals(PackageIdentity.State.PRESENT, result.state)
        assertEquals(UID, result.observed!!.appId)
    }

    @Test
    fun aWorkingLookupIsPresentWithTheCallersAppId() {
        val present = installed(arrayOf(ORIGINAL), arrayOf(ORIGINAL))
        packages.`when`<PackageInfo?> { PackageManagerApis.getPackageInfo(anyString(), anyLong(), anyInt()) }
            .thenReturn(present)

        val result = PackageIdentity.of(PACKAGE, 10)
        assertEquals(PackageIdentity.State.PRESENT, result.state)
        assertEquals(UID, result.observed!!.appId)
        assertEquals(10, result.observed!!.userId)
    }

    @Test
    fun theLookupFlagsAskForSignaturesAndForUninstalledPackages() {
        val flags = PackageIdentity.lookupFlags()
        assertTrue((flags and PackageManager.MATCH_UNINSTALLED_PACKAGES.toLong()) != 0L)
        assertTrue((flags and PackageManager.GET_SIGNING_CERTIFICATES.toLong()) != 0L)
    }

    @Test
    fun anOrdinaryUpdateBySameSignerMatches() {
        val recorded = identityOf(ORIGINAL)
        assertTrue(recorded.matches(observe(installed(arrayOf(ORIGINAL), arrayOf(ORIGINAL)), 0)))
    }

    @Test
    fun aRotatedKeyMatchesThroughItsLineage() {
        val recorded = identityOf(ORIGINAL)
        val rotated = installed(arrayOf(ROTATED), arrayOf(ORIGINAL, ROTATED))
        assertTrue(recorded.matches(observe(rotated, 0)))
    }

    @Test
    fun aForeignSignerDoesNotMatch() {
        val recorded = identityOf(ORIGINAL)
        val foreign = installed(arrayOf(FOREIGN), arrayOf(FOREIGN))
        assertFalse(recorded.matches(observe(foreign, 0)))
    }

    @Test
    fun aMultiSignerPackageComparesAsAnUnorderedSet() {
        val recorded = identityOf(ORIGINAL, SECOND)
        val permuted = installed(arrayOf(SECOND, ORIGINAL), null)
        assertTrue(recorded.matches(observe(permuted, 0)))
    }

    @Test
    fun aMultiSignerPackageWithAForeignSignerDoesNotMatch() {
        // hasMultipleSigners() means no lineage exists, so the whole set is the identity.
        val recorded = identityOf(ORIGINAL, SECOND)
        val replaced = installed(arrayOf(ORIGINAL, FOREIGN), null)
        assertFalse(recorded.matches(observe(replaced, 0)))
    }

    @Test
    fun oneCoSignerOfATwoSignerIdentityIsNotThatIdentity() {
        // Recorded as signed by ORIGINAL and SECOND together. A package signed by ORIGINAL alone,
        // carrying ORIGINAL in its lineage, dropped a signing authority; that is a different
        // installation, not a rotation of the pair.
        val recorded = identityOf(ORIGINAL, SECOND)
        val dropped = installed(arrayOf(ORIGINAL), arrayOf(ORIGINAL))
        assertFalse("a single co-signer matched a two-signer identity", recorded.matches(observe(dropped, 0)))
    }

    @Test
    fun aDifferentAppIdDoesNotMatchEvenWithTheSameSigner() {
        val recorded = identityOf(ORIGINAL)
        val moved = installed(arrayOf(ORIGINAL), arrayOf(ORIGINAL))
        moved.applicationInfo!!.uid = UID + 1
        assertFalse(recorded.matches(observe(moved, 0)))
    }

    @Test
    fun anUnreadableSignatureDoesNotBecomeAnIdentityWithoutSigners() {
        val unreadable = installed(arrayOf(ORIGINAL), arrayOf(ORIGINAL))
        `when`(unreadable.signingInfo!!.apkContentsSigners).thenReturn(arrayOf())
        // The identical reading, through the other entry point, is a failure rather than an answer.
        assertEquals(PackageIdentity.State.LOOKUP_FAILED, PackageIdentity.classify(unreadable, 0).state)

        val identity = PackageIdentity.identityOf(unreadable)
        assertTrue(
            "identityOf produced an identity with no signers: $identity",
            identity == null || identity.signerDigests.isNotEmpty(),
        )
    }

    @Test
    fun twoUsersReportingDifferentSignersIsUnstableRatherThanAReplacement() {
        val first = observe(installed(arrayOf(ORIGINAL), arrayOf(ORIGINAL)), 0)
        val second = observe(installed(arrayOf(FOREIGN), arrayOf(FOREIGN)), 10)
        assertFalse(PackageIdentity.sameInstallation(first, second))
        assertTrue(PackageIdentity.sameInstallation(first, first))
    }

    companion object {
        const val PACKAGE = "eu.darken.porter.probe"
        const val UID = 10123

        val ORIGINAL = Signature("0a0b")
        val ROTATED = Signature("0c0d")
        val FOREIGN = Signature("0e0f")
        val SECOND = Signature("1a1b")

        /** A package installed for one user, signed by [signers] with lineage [history]. */
        fun installed(signers: Array<Signature>, history: Array<Signature>?): PackageInfo {
            val packageInfo = PackageInfo()
            packageInfo.packageName = PACKAGE
            val applicationInfo = ApplicationInfo()
            applicationInfo.uid = UID
            applicationInfo.flags = ApplicationInfo.FLAG_INSTALLED
            packageInfo.applicationInfo = applicationInfo
            val signingInfo = mock(SigningInfo::class.java)
            `when`(signingInfo.hasMultipleSigners()).thenReturn(signers.size > 1)
            `when`(signingInfo.apkContentsSigners).thenReturn(signers)
            `when`(signingInfo.signingCertificateHistory).thenReturn(if (signers.size > 1) null else history)
            packageInfo.signingInfo = signingInfo
            return packageInfo
        }

        fun identityOf(vararg signers: Signature): PackageIdentity.Identity =
            PackageIdentity.identityOf(installed(arrayOf(*signers), arrayOf(*signers)))!!

        fun observe(packageInfo: PackageInfo, userId: Int): PackageIdentity.Observed {
            val result = PackageIdentity.classify(packageInfo, userId)
            assertEquals(PackageIdentity.State.PRESENT, result.state)
            return result.observed!!
        }
    }
}
